Document UPD-24 and UPD-28 (#878)
digital-asset/docs.daml.com: docs/2.10.0/docs/upgrade/smart-contract-upgrades.rst

techniques are listed :ref:`here <upgrades-index>`.

For this reason, we strongly recommend against using Daml Script as an upgrade
dependency for any package going on the ledger. Whenever building the ``main``
package above, you can expect a warning:

.. code:: bash

  > daml build
  ...
  warning while type checking data type Main.MyData:
    This package depends on a datatype <package ID>:Daml.Script.Stable:PartyIdHint from <package ID> (daml-script-lts-stable, 0.0.0) with LF version 1.17.
    
    We do not recommend using datatypes from Daml Script in >= LF1.17 packages, because those datatypes are not upgradeable.
    Upgrade this warning to an error -Werror=upgrade-serialized-daml-script
    Remove this warning with -Wno-upgrade-serialized-daml-script

**Note:** For added safety, you may upgrade these warnings to errors with
``-Werror=upgrade-serialized-daml-script``. We recommend against removing these
warnings with ``-Wno-upgrade-serialized-daml-script``.

If instead ``main`` depends on a datatype in a non-serializable position such
as the type signature of a function, ``main`` can still be upgraded without

> breaking any SCU restrictions. For example:

Reviewer: I'm wondering if we shouldn't drop that part. It may give a false sense of security to the users. There's still the problem that if you upload two different versions of daml-script-lts to the ledger, they may not be compatible. Then again, if we ever break that, we should ear our own dogfood and call it daml-script-lts-v2.

Author: I've been playing with rewriting that section, and couldn't find anything better than a the bolded Note approach that we already have. I don't want to unmention it - I'd rather customers be aware of a workaround, along with its caveats, than that they discover it on their own and think it's a solution with no drawbacks.

in dedicated packages for running tests written in Daml Script. These packages
should not be part of your model and should not be uploaded to the ledger.

Daml Script packages are not guaranteed to be upgradeable across SDK versions.
For this reason, if you depend on a Daml Script datatype in a serializable
position (e.g. the field of a template), you may find your package will rely on
a Daml Script package in a way that can neither be removed nor upgraded to the
next SDK version. At this point, your package and any of its SCU upgrades would
be stuck on that SDK version.

For example, suppose you have a ``main`` package that depends on
``daml-script-lts`` from SDK 2.10.0.

.. code:: yaml

  version: 2.10.0
  name: main
  ...
  dependencies:
  - daml-prim
  - daml-stdlib
  - daml-script-lts
  build-options:
  - --target=1.17

.. code:: daml

  module Main where

  import qualified Daml.Script

  data MyData = MyData
    { field1 : Daml.Script.PartyIdHint
    }

Because ``MyData`` is a serializable datatype, any changes to it must be valid
upgrade changes (e.g. adding a field) for the ``main`` package itself to be
validly upgraded. If SDK 2.10.X were to introduce a change to

> ``daml-script-lts`` that is not a valid upgrade of the ``daml-script-lts`` in

Reviewer 2: It might be worth mentioning that its very very unlikely that we'll do this, for peace of mind of those who have already fallen prey, but up to you. It is somewhat counter to our point

Author: Updated.

    
    It is not recommended that >= LF1.17 packages use datatypes from Daml Script, because those datatypes will not be upgradeable.
    Upgrade this warning to an error -Werror=upgrade-serialized-daml-script
    Disable this warning entirely with -Wno-upgrade-serialized-daml-script

**Note:** For added safety, you may upgrade these warnings to errors with
``-Werror=upgrade-serialized-daml-script``. We recommend against disabling these
warnings with ``-Wno-upgrade-serialized-daml-script``.

If instead ``main`` depended on a datatype in a non-serializable position such
as the type signature of a function, ``main`` could still be upgraded without
breaking any SCU restrictions. For example:

.. code:: daml

  module Main where

  import qualified Daml.Script

  data MyData = MyData
    { field1 : Text
    }

  myFunction : Daml.Script.PartyIdHint -> Bool
  myFunction _ = True

In this case, when changing SDK from 2.10.0 to 2.10.X, the typechecker will
ignore the change of ``Daml.Script.PartyIdHint`` in ``myFunction``, because it
is not in a serializable position. This means ``daml-script-lts`` can be kept

> even though when it is not a valid upgrade from one version to the next.

Reviewer 2: Note that the upload check would still warn in this case though. We warn if daml-script is uploaded at all.
I realise we discussed this before in the context of depending on 1.15 packages.
Its okay to depend on 1.15 if not in a serializable position
Its less okay to depend on daml-script, even if not serializable, if its uploaded.
We should probably just add a caveat about this. It might need rethinking to be a nicer experience.

Reviewer 2: I would say that having a 1.15 type in your template is worse than having a daml-script type though, since we are being careful with script. Up to you if you want to convey that

  ...
  warning while type checking data type Main.MyData:
    This package has LF version 1.17, but it depends on a serializable type <dep1 package ID>:Dep1:D from package <dep 1 package ID> (dep1, 1.0.0) which has LF version 1.15.
    
    It is not recommended that >= LF1.17 packages depend on <= LF1.15 datatypes in places that may be serialized to the ledger, because those datatypes will not be upgradeable.
    Upgrade this warning to an error -Werror=upgrade-serialized-non-upgradeable-dependency
    Disable this warning entirely with -Wno-upgrade-serialized-non-upgradeable-dependency

**Note:** For added safety, you may upgrade these warnings to errors with
``-Werror=upgrade-serialized-non-upgradeable-dependency``. We recommend against disabling these
warnings with ``-Wno-upgrade-serialized-non-upgradeable-dependency``.

.. _upgrade_dont_upload_daml_script:

Avoid Depending on Daml Script Packages
---------------------------------------

We recommend only depending on Daml Script packages such as ``daml-script-lts``
in dedicated packages for running tests written in Daml Script. These packages
should not be part of your model and should not be uploaded to the ledger.

> Daml Script packages are not guaranteed to be upgradeable across SDK versions.
> For this reason, if you depend on a Daml Script datatype in a serializable
> position (e.g. the field of a template), you may find your package will rely on
> a Daml Script package in a way that can neither be removed nor upgraded to the
> next SDK version. At this point, your package and any of its SCU upgrades would
> be stuck on that SDK version.

Reviewer 3: ```suggestion
Daml Script packages are not guaranteed to be upgradeable across SDK versions.
If you depend on a Daml Script datatype in a serializable
position (e.g. the field of a template), your package may rely on
a Daml Script package in a way that can neither be removed nor upgraded to the
next SDK version. Your package and any of its SCU upgrades would
be stuck on that SDK version.
```

For example, suppose you have a ``main`` package that depends on
``daml-script-lts`` from SDK 2.10.0.

.. code:: yaml

  version: 2.10.0
  name: main
  ...
  dependencies:
  - daml-prim
  - daml-stdlib
  - daml-script-lts
  build-options:
  - --target=1.17

.. code:: daml

  module Main where

  import qualified Daml.Script

  data MyData = MyData
    { field1 : Daml.Script.PartyIdHint
    }

> Because ``MyData`` is a serializable datatype, any changes to it must be valid
> upgrade changes (e.g. adding a field) for the ``main`` package itself to be
> validly upgraded. If SDK 2.10.X were to introduce a change to
> ``daml-script-lts`` that is not a valid upgrade of the ``daml-script-lts`` in
> SDK 2.10.0, then ``field1`` will not be upgradeable to the next version of the
> SDK, nor can the field be dropped because the field is used in SCU checks.

Reviewer 3: ```suggestion
Because ``MyData`` is a serializable datatype, any changes to it must be valid
upgrade changes (e.g. adding a field) for the ``main`` package itself to be upgraded. If SDK 2.10.X introduces a change to
``daml-script-lts`` that is not a valid upgrade of the ``daml-script-lts`` in
SDK 2.10.0, then ``field1`` is not upgradeable to the next version of the
SDK, nor can the field be dropped because the field is used in SCU checks.
```

> **Note**: While it is unlikely that an SDK update will break ``daml-script-lts``
> for upgrades, we still strongly recommend against it. In Daml 3.x, Canton may
> disallow uploading ``daml-script-lts`` entirely.

Reviewer 3: ```suggestion
**Note**: While it is unlikely that an SDK update breaks ``daml-script-lts``
for upgrades, we still strongly recommend against it. In Daml 3.x, Canton may
disallow uploading ``daml-script-lts`` entirely.
```

At that point, all future development on ``main``, including future upgrades,
would be locked to SDK 2.10.0. To bump the SDK version, ``main`` would have to
be migrated via a manual upgrade tool with downtime -- existing downtime upgrade
techniques are listed :ref:`here <upgrades-index>`.

For this reason, we strongly recommend against using Daml Script as an upgrade
dependency for any package going on the ledger. Whenever building the ``main``
package above, you can expect a warning:

.. code:: bash

  > daml build
  ...
  warning while type checking data type Main.MyData:
    This package depends on a datatype <package ID>:Daml.Script.Stable:PartyIdHint from <package ID> (daml-script-lts-stable, 0.0.0) with LF version 1.17.
    

>     It is not recommended that >= LF1.17 packages use datatypes from Daml Script, because those datatypes will not be upgradeable.
>     Upgrade this warning to an error -Werror=upgrade-serialized-daml-script
>     Disable this warning entirely with -Wno-upgrade-serialized-daml-script

Reviewer 3: ```suggestion
    We do not recommend using datatypes from Daml Script in >= LF1.17 packages, because those datatypes are not upgradeable.
    Upgrade this warning to an error -Werror=upgrade-serialized-daml-script
    Remove this warning with -Wno-upgrade-serialized-daml-script
```

> **Note:** For added safety, you may upgrade these warnings to errors with
> ``-Werror=upgrade-serialized-daml-script``. We recommend against disabling these
> warnings with ``-Wno-upgrade-serialized-daml-script``.

Reviewer 3: ```suggestion
**Note:** For added safety, you may upgrade these warnings to errors with
``-Werror=upgrade-serialized-daml-script``. We recommend against removing these
warnings with ``-Wno-upgrade-serialized-daml-script``.
```
Again, seems redundant.

> If instead ``main`` depended on a datatype in a non-serializable position such
> as the type signature of a function, ``main`` could still be upgraded without
> breaking any SCU restrictions. For example:

Reviewer 3: ```suggestion
If instead ``main`` depends on a datatype in a non-serializable position such
as the type signature of a function, ``main`` can still be upgraded without
breaking any SCU restrictions. For example:
```

.. code:: daml

  module Main where

  import qualified Daml.Script

  data MyData = MyData
    { field1 : Text
    }

  myFunction : Daml.Script.PartyIdHint -> Bool
  myFunction _ = True

> In this case, when changing SDK from 2.10.0 to 2.10.X, the typechecker will
> ignore the change of ``Daml.Script.PartyIdHint`` in ``myFunction``, because it
> is not in a serializable position. This means ``daml-script-lts`` can be kept
> even though when it is not a valid upgrade from one version to the next.

Reviewer 3: ```suggestion
In this case, when changing SDK from 2.10.0 to 2.10.X, the typechecker
ignores the change to ``Daml.Script.PartyIdHint`` in ``myFunction``, because it
is not in a serializable position. This means ``daml-script-lts`` can be kept
even when it is not a valid upgrade from one version to the next.
```
